File: 20_07_July/0727_42583_다리를_지나는_트럭/Solution.py
from collections import deque

class Truck():

    def __init__(self, weight, time):

        self.weight = weight
        self.time = time


def solution(bridge_length, weight, truck_weights):
    seconds = 0
    totalWeights = 0
    moving = deque()
    trucks = deque(truck_weights)
    
    while moving or trucks:

        if moving and (seconds - moving[0].time) == bridge_length:
            totalWeights -= moving[0].weight
            moving.popleft()
        
        if trucks and (totalWeights + trucks[0]) <= weight:
            totalWeights += trucks[0]
            moving.append(Truck(trucks[0], seconds))
            trucks.popleft()
        
        seconds += 1

    return seconds

print(solution(2, 10, [7, 4, 5, 6]))

File: 20_07_July/0727_42583_다리를_지나는_트럭/Solution.cpp
#include <vector>
#include <cassert>
#include <queue>

using namespace std;

class Truck
{
private:
    int weight_;
    int start_time_;

public:
    Truck(int weight, int time);
    int GetWeight();
    int GetStartTime();

};

Truck::Truck(int weight, int time)
    : weight_(weight), start_time_(time) { }

inline int Truck::GetWeight()
{
    return weight_;
}

inline int Truck::GetStartTime()
{
    return start_time_;
}


int solution(int bridge_length, int weight, vector<int> truck_weights)
{
    int seconds = 0;
    int totalWeights = 0;
    queue<int> truckQueue;
    queue<Truck*> movingQueue;

    for (int w : truck_weights)
    {
        truckQueue.push(w);
    }

    while (truckQueue.empty() == false || movingQueue.empty() == false)
    {
        
        if (movingQueue.empty() == false && (seconds - movingQueue.front()->GetStartTime()) == bridge_length)
        {
            totalWeights -= movingQueue.front()->GetWeight();
            delete movingQueue.front();
            movingQueue.pop();
        }

        if (truckQueue.empty() == false && (totalWeights + truckQueue.front()) <= weight)
        {
            int w = truckQueue.front();
            totalWeights += w;
            movingQueue.push(new Truck(w, seconds));
            truckQueue.pop();
        }

        seconds += 1;
    }
    

    return seconds;
}

void do_test()
{

    assert(solution(2, 10, { 7, 4, 5, 6 }) == 8);
}